src/main/scala/strawman/collection/mutable/ListBuffer.scala

```diff
@@ -33,29 +41,163 @@ class ListBuffer[A]
     aliased = false
   }
 
+  private def ensureUnaliased() = if (aliased) copyElems()
+
   /** Convert to list; avoids copying where possible. */
   def toList = {
     aliased = true
     first
   }
 
+  def clear(): Unit = {
+    first = Nil
+  }
+
   def +=(elem: A) = {
-    if (aliased) copyElems()
-    val last1 = elem :: Nil
-    last match {
-      case last: ::[A] => last.next = last1
-      case _ => first = last1
-    }
+    ensureUnaliased()
+    val last1 = (elem :: Nil).asInstanceOf[::[A]]
+    if (len == 0) first = last1 else last.next = last1
     last = last1
     len += 1
     this
   }
 
+  private def locate(i: Int): Predecessor[A] =
+    if (i == 0) null
+    else if (i == len) last
+    else {
+      assert(i > 0 && i < len)
+      var j = i - 1
+      var p = first
+      while (j > 0) {
+        p = p.tail
+        j -= 1
+      }
+      p.asInstanceOf[Predecessor[A]]
+    }
+
+  private def getNext(p: Predecessor[A]): List[A] =
+    if (p == null) first else p.next
+
+  private def setNext(p: Predecessor[A], nx: List[A]): Unit =
+    if (p == null) first = nx else p.next = nx
+
```

Comment:
Maybe a `swapInPlace` would be good here to abstract over the `update` and `mapInPlace`-type operations?